File: breadth-first-search/breadth-first-search.js
const graph = {};
graph["you"] = ["alice", "bob","claire"];
graph["bob"] = ["anuj", "peggy"];
graph["alice"] = ["peggy"];
graph["claire"] = ["thom", "jonny"];
graph["anuj"] = [];
graph["peggy"] = [];
graph["thom"] = [];
graph["jonny"] = [];

function personIsSeller(name) {
  return name[name.length - 1] === "m";
} 

function search(name){
  let searchQueue = []
  searchQueue.push(name);
  let searched = [];

  while(searchQueue.length){
    let person = searchQueue.shift();
    if (searched.indexOf(person) === -1 ){
      if(personIsSeller(person)){
        console.log(person + " is a mango seller");
        return true;
      }
      else {
      searchQueue = searchQueue.concat(graph[person]);
      searched.push(person);
      }
    }   
    
  }
return false
}


console.log(search("you"));

File: breadth-first-search/breadth_first_search.rb
graph = {}

graph["you"] = %w(alice bob claire)
graph["bob"] = %w(anuj peggy)
graph["alice"] = %w(peggy)
graph["claire"] = %w(thom jonny)
graph["anuj"] = []
graph["peggy"] = []
graph["thom"] = []
graph["jonny"] = []



def person_is_seller(name)
   name[-1] === "m"
end


def search(name, g)
  search_queue = []
  search_queue.push(name)
  searched = []
 
  while search_queue.length 
    person = search_queue.shift()
    unless searched.include?(person)
      if person_is_seller(person)
        puts("#{person} is a mango seller")
        return true
      else  
        search_queue += g[person];
        searched.push(person)
      end
    end    
  end
  return false
end

puts(search("you", graph))

